docs: choose the state structure (#1114)
deephaven/deephaven-plugins: plugins/ui/docs/managing-state/choose-the-state-structure.md

# Choose the State Structure

> Structuring state effectively can be the difference between a component that is easy to modify and debug, and one that is a persistent source of bugs. Here are some tips to consider when organizing state.

Reviewer: ```suggestion
Effectively structuring state can be the difference between a component that is easy to modify and debug and one that is a persistent source of bugs. Here are some tips to consider when organizing state.
```

1. **Group related state.** If you always update two or more state variables at the same time, consider merging them into a single state variable.
2. **Avoid contradictions in state.** When the state is structured in a way that several pieces of state may contradict and “disagree” with each other, you leave room for mistakes. Try to avoid this.
3. **Avoid redundant state.** If you can calculate some information from the component’s props or its existing state variables during rendering, you should not put that information into that component’s state.
4. **Avoid duplication in state.** When the same data is duplicated between multiple state variables, or within nested objects, it is difficult to keep them in sync. Reduce duplication when you can.
5. **Avoid deeply nested state.** Deeply hierarchical state is not very convenient to update. When possible, prefer to structure state in a flat way.

> The purpose of these principles is to simplify state updates and minimize errors. By eliminating redundant and duplicate data from the state, you can ensure consistency across all its pieces. This approach is akin to how a database engineer might "normalize" a database structure to minimize bugs.

Reviewer: ```suggestion
These principles aim to simplify state updates and minimize errors. By eliminating redundant and duplicate data from the state, you can ensure consistency across all its pieces. This approach is akin to how a database engineer might "normalize" a database structure to minimize bugs.
```

Since `is_sending` and `is_sent` should never be `True` at the same time, it is better to replace them with a single status state variable that can take one of three valid states: `typing` (initial), `sending`, and `sent`:

```python
from deephaven import ui
import threading

@ui.component
def feedback_form():
    text, set_text = ui.use_state("")
    status, set_status = ui.use_state("typing")

    def finish_submit():
        set_status("sent")

    def handle_submit():
        set_status("sending")
        threading.Timer(5, finish_submit).start()

    is_sending = status == "sending"
    is_sent = status == "sent"

    if is_sent:
        return ui.heading("Thanks for the feedback!")

    return ui.form(
        ui.text("Do you have any feedback?"),
        ui.text_area(value=text, on_change=set_text, is_disabled=is_sending),
        ui.button("Send", type="submit"),
        ui.text("Sending...") if is_sending else None,
        on_submit=handle_submit,
    )

feedback_form_example = feedback_form()
```

You can still declare some constants for readability:

```python
is_sending = status == "sending"
is_sent = status == "sent"
```

> But they are not state variables, so you do not need to worry about them getting out of sync with each other.

Reviewer: ```suggestion
However, they are not state variables, so you do not need to worry about them getting out of sync with each other.
```

    ]

name_input_example = name_input()
```

Here, `full_name` is not a state variable. Instead, it is calculated during render:

```python
full_name = f"{first_name} {last_name}"
```

Therefore, the change handlers can update it without any special actions. Calling `set_first_name` or `set_last_name` triggers a re-render, and the `full_name` will be recalculated using the updated data.

## Avoid duplication in state

This component lets you choose a single snack out of several:

```python
from deephaven import ui

initial_items = [
    {"title": "peanuts", "id": 0},
    {"title": "chips", "id": 1},
    {"title": "apples", "id": 2},
]

@ui.component
def snacks():
    items, set_items = ui.use_state(initial_items)
    selected_item, set_selected_item = ui.use_state(items[0])

    def snack_row(item):
        return ui.flex(
            ui.text(item["title"]),
            ui.button("Choose", on_press=lambda: set_selected_item(item)),
        )

    return [
        ui.heading("What is your favorite snack?"),
        [snack_row(item) for item in items],
        ui.text(f"You picked {selected_item['title']}"),
    ]

snacks_example = snacks()
```

> The selected item is currently stored as an dictionary in the `selected_item` state variable. However, this approach is not ideal because the `selected_item` contains the same object as one of the items in the `items` list. This results in duplicated information about the item in two places.

Reviewer: ```suggestion
The selected item is currently stored as a dictionary in the `selected_item` state variable. However, this approach is not ideal because the `selected_item` contains the same object as one of the items in the `items` list. This results in duplicated information about the item in two places.
```
